provisionersdk/proto/provisioner.proto

```diff
@@ -168,96 +168,111 @@ message Resource {
+message ParseRequest {
+}
 
-        string provisioner_log_level = 4;
-    }
+// ParseComplete indicates a request to parse completed.
+message ParseComplete {
+    string error = 1;
+    repeated TemplateVariable template_variables = 2;
+	bytes readme = 3;
+}
 
-    message Plan {
-        reserved 2;
+// PlanRequest asks the provisioner to plan what resources & parameters it will create
+message PlanRequest {
+	Metadata metadata = 1;
+    repeated RichParameterValue rich_parameter_values = 2;
+    repeated VariableValue variable_values = 3;
+    repeated GitAuthProvider git_auth_providers = 4;
+}
 
-        Config config = 1;
-        repeated RichParameterValue rich_parameter_values = 3;
-        repeated VariableValue variable_values = 4;
-        repeated GitAuthProvider git_auth_providers = 5;
-    }
+// PlanComplete indicates a request to plan completed.
+message PlanComplete {
+    string error = 1;
+    repeated Resource resources = 2;
+    repeated RichParameter parameters = 3;
+    repeated string git_auth_providers = 4;
+}
+
+// ApplyRequest asks the provisioner to apply the changes.  Apply MUST be preceded by a successful plan request/response
+// in the same Session.  The plan data is not transmitted over the wire and is cached by the provisioner in the Session.
+message ApplyRequest {
+	Metadata metadata = 1;
+}
 
-    message Apply {
+// ApplyComplete indicates a request to apply completed.
+message ApplyComplete {
+    bytes state = 1;
+    string error = 2;
+    repeated Resource resources = 3;
+    repeated RichParameter parameters = 4;
+    repeated string git_auth_providers = 5;
+}
+
+// CancelRequest requests that the previous request be canceled gracefully.
+message CancelRequest {}
+
+message Request {
+    oneof type {
         Config config = 1;
-        bytes plan = 2;
+        ParseRequest parse = 2;
+        PlanRequest plan = 3;
+        ApplyRequest apply = 4;
+        CancelRequest cancel = 5;
     }
+}
 
-    message Cancel {}
-    message Request {
-        oneof type {
-            Plan plan = 1;
-            Apply apply = 2;
-            Cancel cancel = 3;
-        }
-    }
-    message Complete {
-        bytes state = 1;
-        string error = 2;
-        repeated Resource resources = 3;
-        repeated RichParameter parameters = 4;
-        repeated string git_auth_providers = 5;
-        bytes plan = 6;
-    }
-    message Response {
-        oneof type {
-            Log log = 1;
-            Complete complete = 2;
-        }
+message Response {
+    oneof type {
+        Log log = 1;
+        ParseComplete parse = 2;
+        PlanComplete plan = 3;
+        ApplyComplete apply = 4;
     }
 }
 
 service Provisioner {
-    rpc Parse(Parse.Request) returns (stream Parse.Response);
-    rpc Provision(stream Provision.Request) returns (stream Provision.Response);
+    // Session represents provisioning a single template import or workspace.  The daemon always sends Config followed
```

Comment:
I don't believe we need to update any docs --- this interface is entirely within the provisioner daemon for now.

One thing it does affect is the format of messages on the Echo provisioner, which _technically_ you could install in a real system.  But, its only for testing and we don't publicly document it today (nor do I think we should!).

Reply:
Actually, I was thinking about updating this [page](https://coder.com/docs/v2/latest/about/architecture) accordingly, but it can be performed in a followup.

Reply:
Yeah, this PR doesn't change the architecture, but it is true that a subsequent PR will allow the provisioner to be separate from the provisioner daemon, which would be an architecture change.